Libs/vtkSegmentationCore/vtkSegmentation.cxx

```diff
@@ -1573,13 +1573,24 @@ void vtkSegmentation::SeparateSegmentLabelmap(std::string segmentId)
     thresholdErase->SetInValue(0);
     thresholdErase->ReplaceOutOff();
     thresholdErase->Update();
+
+    // Although the source labelmap will be modified, onSourceRepresentationModified would invalidate the non-source representations.
+    // This prevents onSourceRepresentationModified from being called.
+    bool wasSourceRepresentationModifiedEnabled = this->SetSourceRepresentationModifiedEnabled(false);
     labelmap->ShallowCopy(thresholdErase->GetOutput());
+    this->SetSourceRepresentationModifiedEnabled(wasSourceRepresentationModifiedEnabled);
     }
   segment->SetLabelValue(DEFAULT_LABEL_VALUE);
 
+  std::string sourceRepresentationName = this->GetSourceRepresentationName();
+  if (strcmp(sourceRepresentationName.c_str(), vtkSegmentationConverter::GetBinaryLabelmapRepresentationName()) == 0)
+    {
+    // We were blocking onSourceRepresentationModified, however the source labelmap was modified so we need to manually invoke the event.
+    this->InvokeEvent(vtkSegmentation::SourceRepresentationModified, labelmap);
+    }
+  this->InvokeEvent(vtkSegmentation::RepresentationModified, labelmap);
```

Comment:
Could you please check if these 3 events (vtkSegmentation::SourceRepresentationModified, vtkSegmentation::RepresentationModified, this->Modified()) do not result in performing any expensive operation (e.g., regeneration of closed surface representation) multiple times? (most probably not, just to make sure)

Reply:
Summary of event observations:
- SourceRepresentationModified:
  - updateWidgetFromMRML in qSlicerSegmentationsModuleWidget
  - updateSliceRotateWarningButtonVisibility in qMRMLSegmentEditorWidget.
  - populateRepresentationsList in qMRMLSegmentationRepresentationsListView.
  - Segmentation history
  - Update storage node supported write file types.
  - Propagate event through segmentation node.

- RepresentationModified:
  - Displayable managers.
  - Update StorableModifiedTime.
  - Propagate event through segmentation node..

- ModifiedEvent:
  - Not observed anywhere